### ISSUE TITLE
Implement Local Open-Source Voice Activity Detection (VAD)

### ISSUE THREAD

**Issue by csmangum** [csmangum/opusagent#134]
Implement a local, open-source Voice Activity Detection (VAD) solution to replace or complement the current OpenAI Realtime API VAD. This will provide better control over speech detection, reduce latency, and allow for offline operation.

## 🔍 **Current State**
- ✅ VAD configuration infrastructure is in place
- ✅ Environment variable controls (`VAD_ENABLED`, `TUI_VAD_ENABLED`)
- ✅ Bridge architecture supports VAD event handling
- ✅ AudioCodes message models for VAD events
- ✅ Comprehensive testing framework
- ❌ No actual local VAD implementation

## 🎯 **Objective**
Implement a high-quality, low-latency local VAD solution that:
1. Processes audio streams in real-time
2. Detects speech start/stop events accurately
3. Integrates seamlessly with existing infrastructure
4. Provides configurable sensitivity and parameters
5. Supports multiple audio formats and sample rates
6. Maintains high performance with low CPU usage

### Option 1: **Silero VAD** (Recommended)
- **Pros**: State-of-the-art neural network-based, very accurate, actively maintained
- **Cons**: Requires PyTorch, slightly higher CPU usage
- **GitHub**: https://github.com/snakers4/silero-vad
- **Performance**: 95%+ accuracy, ~1ms latency
- **Usage**: Simple API, supports streaming

### Option 2: **WebRTC VAD**
- **Pros**: Battle-tested, very fast, minimal CPU usage
- **Cons**: Less accurate than neural approaches, older algorithm
- **Library**: `webrtcvad-wheels` or `py-webrtcvad`
- **Performance**: 85-90% accuracy, <1ms latency
- **Usage**: Industry standard, used in many applications

### Option 3: **PyTorch Audio VAD**
- **Pros**: Part of PyTorch ecosystem, good accuracy
- **Cons**: Heavier dependency, requires PyTorch
- **Library**: `torchaudio.functional.vad`
- **Performance**: 90-95% accuracy, ~2ms latency

### Option 4: **OpenVAD**
- **Pros**: Lightweight, customizable
- **Cons**: Less mature, limited documentation
- **Performance**: 80-85% accuracy, varies

### Phase 1: Core VAD Implementation
1. **Create VAD Module** (`opusagent/vad/`)
   - `base_vad.py` - Abstract VAD interface
   - `silero_vad.py` - Silero VAD implementation
   - `webrtc_vad.py` - WebRTC VAD implementation
   - `vad_factory.py` - Factory for VAD selection

2. **VAD Configuration**
   - Add VAD-specific environment variables
   - Support multiple VAD backends
   - Configurable sensitivity and parameters

3. **Audio Processing Pipeline**
   - Real-time audio buffering
   - Format conversion utilities
   - Sample rate handling

### Phase 2: Integration
1. **Bridge Integration**
   - Modify `AudioStreamHandler` to use local VAD
   - Replace OpenAI VAD events with local detection
   - Maintain backward compatibility

2. **Event Generation**
   - Generate `speech_started` events locally
   - Generate `speech_stopped` events locally
   - Generate `speech_committed` events locally

### Phase 3: Advanced Features
1. **Adaptive VAD**
   - Noise level adaptation
   - Dynamic sensitivity adjustment
   - Context-aware detection

2. **Performance Optimization**
   - Multi-threading for real-time processing
   - Memory pool management
   - CPU usage monitoring

### Phase 4: Testing & Validation
1. **Comprehensive Testing**
   - Unit tests for VAD modules
   - Integration tests with audio streams
   - Performance benchmarks

2. **Quality Assurance**
   - Accuracy testing with labeled datasets
   - Latency measurements
   - Resource usage profiling

### VAD Module Structure
```
opusagent/vad/
├── __init__.py
├── base_vad.py           # Abstract VAD interface
├── silero_vad.py         # Silero VAD implementation
├── webrtc_vad.py         # WebRTC VAD implementation
├── vad_factory.py        # VAD backend factory
├── audio_processor.py    # Audio processing utilities
├── vad_config.py         # VAD configuration management
└── utils.py              # Utility functions
```

### Configuration Variables
```bash
# VAD Backend Selection
VAD_BACKEND=silero              # silero, webrtc, pytorch
VAD_MODEL_PATH=models/silero_vad.onnx
VAD_CONFIDENCE_THRESHOLD=0.5    # 0.0 to 1.0

# Audio Processing
VAD_SAMPLE_RATE=16000          # Audio sample rate
VAD_CHUNK_SIZE=512             # Audio chunk size
VAD_OVERLAP_SIZE=256           # Overlap between chunks

# Detection Parameters
VAD_MIN_SPEECH_DURATION=0.1    # Minimum speech duration (seconds)
VAD_MIN_SILENCE_DURATION=0.5   # Minimum silence duration (seconds)
VAD_SPEECH_PAD_MS=100          # Padding around speech (milliseconds)

# Performance
VAD_ENABLE_GPU=false           # Use GPU acceleration if available
VAD_NUM_THREADS=2              # Number of processing threads
VAD_BUFFER_SIZE=8192           # Audio buffer size
```

### VAD Interface
```python
class BaseVAD(ABC):
    """Abstract base class for Voice Activity Detection."""
    
    @abstractmethod
    async def initialize(self, config: VADConfig) -> None:
        """Initialize the VAD system."""
        pass
    
    @abstractmethod
    async def process_audio(self, audio_data: bytes) -> VADResult:
        """Process audio data and return VAD result."""
        pass
    
    @abstractmethod
    async def reset(self) -> None:
        """Reset VAD state."""
        pass
    
    @abstractmethod
    async def cleanup(self) -> None:
        """Cleanup VAD resources."""
        pass
```

### Integration Points
1. **AudioStreamHandler**: Integrate VAD processing
2. **RealtimeHandler**: Generate VAD events
3. **EventRouter**: Route local VAD events
4. **Configuration**: Load VAD settings

### Unit Tests
- [ ] VAD module initialization
- [ ] Audio processing pipeline
- [ ] Event generation accuracy
- [ ] Configuration loading
- [ ] Resource cleanup

### Integration Tests
- [ ] Real-time audio processing
- [ ] Bridge event generation
- [ ] Performance under load
- [ ] Multiple audio formats
- [ ] Concurrent sessions

### Performance Tests
- [ ] Latency measurements
- [ ] CPU usage profiling
- [ ] Memory usage tracking
- [ ] Accuracy benchmarks
- [ ] Stress testing

### Validation Tests
- [ ] Compare with OpenAI VAD
- [ ] Test with different speakers
- [ ] Noise robustness testing
- [ ] Edge case handling
- [ ] Long-duration testing

### Accuracy Metrics
- **Speech Detection Accuracy**: >95%
- **False Positive Rate**: <5%
- **False Negative Rate**: <3%
- **Timing Accuracy**: ±50ms

### Performance Metrics
- **Processing Latency**: <10ms
- **CPU Usage**: <5% per stream
- **Memory Usage**: <50MB per stream
- **Real-time Factor**: >10x

### Quality Metrics
- **Noise Robustness**: Works with SNR >10dB
- **Speaker Independence**: Works across demographics
- **Language Independence**: Works with multiple languages
- **Format Support**: PCM16, μ-law, a-law

### Phase 1: Foundation (Week 1-2)
- [ ] Create VAD module structure
- [ ] Implement Silero VAD backend
- [ ] Add basic configuration
- [ ] Create unit tests

### Phase 2: Integration (Week 3-4)
- [ ] Integrate with AudioStreamHandler
- [ ] Replace OpenAI VAD events
- [ ] Add WebRTC VAD backend
- [ ] Integration testing

### Phase 3: Enhancement (Week 5-6)
- [ ] Add adaptive features
- [ ] Performance optimization
- [ ] Advanced configuration
- [ ] Comprehensive testing

### Phase 4: Validation (Week 7-8)
- [ ] Accuracy validation
- [ ] Performance benchmarking
- [ ] User acceptance testing
- [ ] Documentation completion

### Required
```python
# Core dependencies
torch>=1.9.0                    # For Silero VAD
torchaudio>=0.9.0               # Audio processing
numpy>=1.21.0                   # Numerical operations
scipy>=1.7.0                    # Signal processing

# Optional (for WebRTC VAD)
webrtcvad-wheels>=2.0.10        # WebRTC VAD
```

### Optional
```python
# Performance optimization
numba>=0.54.0                   # JIT compilation
librosa>=0.8.0                  # Audio analysis
soundfile>=0.10.0               # Audio file I/O
```

### Functional Requirements
- [ ] Local VAD processing works in real-time
- [ ] Generates accurate speech start/stop events
- [ ] Configurable via environment variables
- [ ] Supports multiple VAD backends
- [ ] Integrates with existing bridge architecture
- [ ] Maintains backward compatibility

### Performance Requirements
- [ ] Processing latency < 10ms
- [ ] CPU usage < 5% per stream
- [ ] Memory usage < 50MB per stream
- [ ] Accuracy > 95% on standard datasets
- [ ] Handles concurrent sessions efficiently

### Quality Requirements
- [ ] Comprehensive test coverage (>90%)
- [ ] Detailed documentation
- [ ] Performance benchmarks
- [ ] Migration guide from OpenAI VAD
- [ ] Troubleshooting documentation

### Technical Documentation
- [ ] VAD module API documentation
- [ ] Configuration guide
- [ ] Performance tuning guide
- [ ] Architecture overview
- [ ] Integration examples

### User Documentation
- [ ] Setup and installation guide
- [ ] Configuration examples
- [ ] Troubleshooting guide
- [ ] Migration from OpenAI VAD
- [ ] Best practices

### Technical Risks
- **Latency Issues**: Implement efficient buffering and processing
- **Accuracy Problems**: Use multiple VAD backends and validation
- **Resource Usage**: Profile and optimize performance
- **Integration Complexity**: Maintain clear interfaces

### Operational Risks
- **Model Updates**: Version control for VAD models
- **Dependency Management**: Pin versions and test thoroughly
- **Configuration Complexity**: Provide sensible defaults
- **Backward Compatibility**: Maintain existing API

### Advanced Features
- [ ] Multi-speaker VAD
- [ ] Emotion detection integration
- [ ] Custom model training
- [ ] Edge device deployment
- [ ] Federated learning support

### Performance Improvements
- [ ] Hardware acceleration (GPU/TPU)
- [ ] Model quantization
- [ ] Streaming optimization
- [ ] Edge computing support

---

**Note**: This issue builds upon the existing VAD infrastructure implemented in the bridge architecture. The configuration and event handling systems are already in place, making this implementation primarily focused on the actual VAD processing logic.
